File: backend/src/main/java/com/codeloom/backend/config/JdbcConfig.java
package com.codeloom.backend.config;

import com.codeloom.backend.converter.*;
import com.codeloom.backend.model.*;
import com.codeloom.common.SubmissionStatus;
import java.sql.JDBCType;
import java.sql.SQLException;
import java.util.List;
import lombok.RequiredArgsConstructor;
import org.postgresql.util.PGobject;
import org.springframework.context.annotation.Configuration;
import org.springframework.data.convert.*;
import org.springframework.data.jdbc.core.mapping.JdbcValue;
import org.springframework.data.jdbc.repository.config.AbstractJdbcConfiguration;
import tools.jackson.databind.ObjectMapper;

@Configuration
@RequiredArgsConstructor
public class JdbcConfig extends AbstractJdbcConfiguration {
    private final ObjectMapper objectMapper;

    @Override
    protected List<?> userConverters() {
        return List.of(
                new ExamplesRead(objectMapper),
                new ExamplesWrite(objectMapper),
                new ConstraintsRead(objectMapper),
                new ConstraintsWrite(objectMapper),
                new DifficultyWrite(),
                new StatusWrite());
    }

    @ReadingConverter
    static class ExamplesRead extends AbstractJsonReadingConverter<ProblemExamples> {
        ExamplesRead(ObjectMapper m) {
            super(m, ProblemExamples.class);
        }
    }

    @WritingConverter
    static class ExamplesWrite extends AbstractJsonWritingConverter<ProblemExamples> {
        ExamplesWrite(ObjectMapper m) {
            super(m);
        }
    }

    @ReadingConverter
    static class ConstraintsRead extends AbstractJsonReadingConverter<ProblemConstraints> {
        ConstraintsRead(ObjectMapper m) {
            super(m, ProblemConstraints.class);
        }
    }

    @WritingConverter
    static class ConstraintsWrite extends AbstractJsonWritingConverter<ProblemConstraints> {
        ConstraintsWrite(ObjectMapper m) {
            super(m);
        }
    }

    @WritingConverter
    static class DifficultyWrite
            implements org.springframework.core.convert.converter.Converter<ProblemDifficulty, JdbcValue> {
        public JdbcValue convert(ProblemDifficulty s) {
            return enumValue(s.name(), "problem_difficulty");
        }
    }

    @WritingConverter
    static class StatusWrite
            implements org.springframework.core.convert.converter.Converter<SubmissionStatus, JdbcValue> {
        public JdbcValue convert(SubmissionStatus s) {
            return JdbcValue.of(s.name(), JDBCType.VARCHAR);
        }
    }

    private static JdbcValue enumValue(String value, String type) {
        try {
            var object = new PGobject();
            object.setType(type);
            object.setValue(value);
            return JdbcValue.of(object, JDBCType.OTHER);
        } catch (SQLException exception) {
            throw new IllegalArgumentException(exception);
        }
    }
}

File: backend/src/main/kotlin/com/codeloom/backend/config/JdbcConfig.kt
package com.codeloom.backend.config

import com.codeloom.backend.converter.AbstractJsonReadingConverter
import com.codeloom.backend.converter.AbstractJsonWritingConverter
import com.codeloom.backend.model.ProblemConstraints
import com.codeloom.backend.model.ProblemDifficulty
import com.codeloom.backend.model.ProblemExamples
import com.fasterxml.jackson.databind.ObjectMapper
import org.springframework.context.annotation.Configuration
import org.springframework.core.convert.converter.Converter
import org.springframework.data.convert.ReadingConverter
import org.springframework.data.convert.WritingConverter
import org.springframework.data.jdbc.core.mapping.JdbcValue
import org.springframework.data.jdbc.repository.config.AbstractJdbcConfiguration
import java.sql.JDBCType

@Configuration
class JdbcConfig(
    private val objectMapper: ObjectMapper,
) : AbstractJdbcConfiguration() {

    override fun userConverters(): List<*> {
        return listOf(
            // Problem Examples
            @ReadingConverter
            object : AbstractJsonReadingConverter<ProblemExamples>(objectMapper, ProblemExamples::class.java) {},

            @WritingConverter
            object : AbstractJsonWritingConverter<ProblemExamples>(objectMapper, ProblemExamples::class.java) {},

            // Problem constraints
            @ReadingConverter
            object : AbstractJsonReadingConverter<ProblemConstraints>(objectMapper, ProblemConstraints::class.java) {},

            @WritingConverter
            object : AbstractJsonWritingConverter<ProblemConstraints>(objectMapper, ProblemConstraints::class.java) {},

            // ProblemDifficulty to PostgreSQL enum
            @WritingConverter
            object : Converter<ProblemDifficulty, JdbcValue> {
                override fun convert(source: ProblemDifficulty) = JdbcValue.of(source, JDBCType.OTHER)
            },
        )
    }
}